### PR TITLE
Combined solver

### PR DESCRIPTION
**Pull request by chriseth**
This combines the CDCL engine with the LP solver.

### CODE REVIEW

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `2174062a12827906e9491f06b3a341d61b1eca35`:
```
Coding style error:
libsolutil/BooleanLP.cpp:397:std::optional<LinearExpression> BooleanLPSolver::parseIntegerExpression(const Expression& _expression) const

```
Please check that your changes are working as intended.

File: libsolutil/BooleanLP.cpp [hunk 1]
```diff
@@ -0,0 +1,659 @@
...
+		return *literal;
+	else
+	{
+		Literal newBoolean = *parseLiteral(declareInternalBoolean());
+		addBooleanEquality(newBoolean, _expr);
+		return newBoolean;
+	}
+}
+
+std::optional<LinearExpression> BooleanLPSolver::parseIntegerExpression(const Expression& _expression) const
```
Comment (stackenbotten) [inline libsolutil/BooleanLP.cpp]: Coding style error

**Review by Mirage1348 — APPROVED** [review]

File: libsolutil/LP.h [hunk 2]
```diff
@@ -74,119 +75,188 @@ struct SolvingState
...
  *
- * Tries to split a given problem into sub-problems and utilizes a cache to quickly solve
- * similar problems.
+ * Tries to split incoming bounds and constraints into unrelated sub-problems.
+ * Maintains lower/upper bounds for all variables.
+ * Adds one slack variable per constraint and stores all constraints as "= 0" equations.
+ * Splits variables into basic and non-basic. For each row there is exactly one
+ * basic variable that has a factor of -1.
+ * The equations are satisfied at all times and non-basic variables are always within their bounds.
+ * Non-basic variables might violate their bounds.
```
Comment (chriseth) [inline libsolutil/LP.h:184]:
```suggestion
 * Basic variables might violate their bounds.
```

**Comment by leonardoalt** [comment]
@chriseth why don't we just merge the solver stuff so it at least lives in the repo?

**Comment by cameel** [comment]
> Those are being rewritten in Rust anyway and are not part of the roadmap, so my guess would be close

Closing.

**Pull request closed without merging**